main.py

```diff
@@ -102,7 +115,42 @@ def parse_config(config_file):
 def ip_to_int(network):
     """ Takes an IP and returns the unsigned integer encoding of the address """
     return struct.unpack('=L', socket.inet_aton(network))[0]
-    
+
+
+def enrich_event(event):
+    """ Takes the raw event data and enriches by adding process tree metadata """
+    proctree_enriched = []
+    error = ""
+    try:
+        proc = psutil.Process(event.pid)
+        proctree = utils.crawl_process_tree(proc)
+        proctree_enriched = list({"pid": p.pid, "cmdline": " ".join(p.cmdline()), "username":  p.username()} for p in proctree)
+    except Exception as e:
+        error=str(e)
+    return(
+        {"timestamp": datetime.utcnow().isoformat() + 'Z',
+        "pid": event.pid,
+        "proctree": proctree_enriched,
+        "daddr": socket.inet_ntoa(struct.pack('<L', event.daddr)),
+        "port": event.dport,
+        "error": error})
+
+def print_enriched_event(b, out, cpu, data, size):
+    """ A callback for printing enriched event metadata, should be
+    passed as a partial to the callback registering function as
+    `partial(print_enriched_event, b, out)` where `b` is the bpf
+    interface that's being polled and `out` is the output writer
+    (e.g. `sys.stdout`)
+
+    The remaining three arguments (`cpu`, `data` and `size`) are
+    required for the callback, but only `data` is used to pull the
+    event out.
+    """
+
+    event = b["events"].event(data)
+    print >> out, json.dumps(enrich_event(event))
+    out.flush()
```

Comment:
why is flushing the buffer necessary here?

Reply:
This is a continuation of the previous behaviour - when writing to a fifo we found that the OS was just buffering the output indefinitely for some reason... I'd prefer to keep it in for this PR because it's quite a big refactor, but we can play around with removing it once we've got the itests written.